File: automations/dim_weight.py
"""DIM (volumetric) weight calculator for air freight.

Standard air freight DIM factor = 6000 (cm → kg).
Chargeable weight = max(actual_weight, volumetric_weight).
"""


def calculate_chargeable_weight_kg(
    actual_weight_kg: float,
    length_cm: float,
    width_cm: float,
    height_cm: float,
    dim_factor: int = 6000,
) -> float:
    """Return max(actual_weight, volumetric_weight)."""
    volumetric = (length_cm * width_cm * height_cm) / dim_factor
    return max(actual_weight_kg, volumetric)


def total_chargeable_weight(pieces: list, dim_factor: int = 6000) -> float:
    """Sum chargeable weight across all piece lines (count × chargeable per piece)."""
    total = 0.0
    for p in pieces:
        count = p.get("count") or 1
        actual = p.get("weight_kg") or 0
        l = p.get("length_cm") or 0
        w = p.get("width_cm") or 0
        h = p.get("height_cm") or 0
        if l and w and h:
            cw = calculate_chargeable_weight_kg(actual, l, w, h, dim_factor)
        else:
            cw = actual
        total += cw * count
    return round(total, 2)

File: dashboard/src/lib/dim-weight.ts
/**
 * DIM (volumetric) weight calculator for air freight.
 *
 * Standard air freight DIM factor = 6000 (cm → kg).
 * Chargeable weight = max(actual_weight, volumetric_weight).
 */

export function calculateChargeableWeight(
  actualWeightKg: number,
  lengthCm: number,
  widthCm: number,
  heightCm: number,
  dimFactor = 6000
): number {
  const volumetric = (lengthCm * widthCm * heightCm) / dimFactor;
  return Math.max(actualWeightKg, volumetric);
}

export interface PieceInput {
  count?: number | null;
  weight_kg?: number | null;
  length_cm?: number | null;
  width_cm?: number | null;
  height_cm?: number | null;
}

export function totalChargeableWeight(
  pieces: PieceInput[],
  dimFactor = 6000
): number {
  let total = 0;
  for (const p of pieces) {
    const count = p.count ?? 1;
    const actual = p.weight_kg ?? 0;
    const l = p.length_cm ?? 0;
    const w = p.width_cm ?? 0;
    const h = p.height_cm ?? 0;
    const cw =
      l && w && h
        ? calculateChargeableWeight(actual, l, w, h, dimFactor)
        : actual;
    total += cw * count;
  }
  return Math.round(total * 100) / 100;
}
